File: 0120-triangle/0120-triangle.java
class Solution {
    Integer dp[][];
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        dp = new Integer[n][n];
        return solve(triangle, 0, 0);
    }
    public int solve(List<List<Integer>> list, int prevIdx, int row){
        if(row == list.size()-1){
            return list.get(row).get(prevIdx);
        }
        if(dp[row][prevIdx] != null) return dp[row][prevIdx];
        int ans1 = list.get(row).get(prevIdx) + solve(list, prevIdx, row+1);
        int ans2 = list.get(row).get(prevIdx) + solve(list, prevIdx+1, row+1);
        return dp[row][prevIdx] = Math.min(ans1, ans2);
    }
}

File: 0120-triangle/0120-triangle.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int dp[201][201];

    int minimumTotal(vector<vector<int>>& triangle) {
        for(int i = 0; i < 201; i++)
            for(int j = 0; j < 201; j++)
                dp[i][j] = 1001;
        return solve(triangle, 0, 0);
    }

    int solve(vector<vector<int>>& triangle, int i, int j) {
        if(i == triangle.size() - 1)
            return triangle[i][j];
        if(dp[i][j] != 1001)
            return dp[i][j];
        int down = solve(triangle, i + 1, j);
        int diagonal = solve(triangle, i + 1, j + 1);
        return dp[i][j] = triangle[i][j] + min(down, diagonal);
    }
};
